<a href="https://colab.research.google.com/github/tomanizer/stats_in_10_minutes/blob/master/Binomial_Distribution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Binomial distribution

The binomial distribution is a *discrete probability distribution*.
The sample space consists of only two outcomes.

Hence it is suitable for analysing problems which can be framed as *yes/no*-questions or problems which can be interpreted as coin toss.
The two outcomes are often defined as success/fail.

A single success/fail experiment (such as a single toin coss) is called a *Bernoulli trial*. And a sequence of the outcomes such trials is called a *Bernoulli process*.

By convention the success-probability is denoted by $p$ and the number of experiments as $n$.

A *sampling distribution* describes the probability distribution of a given random-sample-based statistic.

A Bernoulli (sampling) distribution, describes the probability distribution of a Bernoulli process with a given success-probability $p$ and $n$ experiments in each sample.

We implement each element of the *Binomial distribution* step by step in Python.

### Bernoulli Trial

In [0]:
from random import random

def bernoulli_trial(p):
  """A single Bernoulli experiment with an outcome of True or False.

  Parameters:
  p (float): probability of True

  Return:
  bool: success or fail 
  """
  x = random() # python returns a random number between 0 and 1
  if x <= p:
    return True # success
  else:
    return False # fail

Perform one *Bernoulli trial* for a coin toss:

In [110]:
p = 0.5

bernoulli_trial(p=p)

True

### Bernoulli Process

In [0]:
def bernoulli_process(n, p):
  """A series of Bernoulli experiments

  Parameters:
  n (int): number of experiments
  p (float): probability of success

  Returns:
  list: series of success and fails
  """
  return [bernoulli_trial(p=p) for i in range(n)]

Create a *Bernoulli process* with a length of 10 experiments.

In [112]:
n = 10

bernoulli_process(n=n, p=p)

[True, False, True, True, False, False, False, True, False, False]

### Bernoulli Sample Distribution

In [0]:
def bernoulli_sample_distribution(trials, n, p):
  """Perform a number of trials with n experiments with success probability p.

  Parameters:
  trials (int): number of samples
  n (int): number of experiments in each sample
  p (float): success probability

  Returns:
  list of lists: samples of bernoulli processes
  """
  return [bernoulli_process(n=n, p=p) for i in range(trials)]

Create a *Bernoullie sample distribution* by perfoming 15 trials of lenth n with a success probability p.

In [119]:
trials = 15

simulation1 = bernoulli_sample_distribution(trials=trials, p=p, n=n)
simulation1

[[True, True, True, True, False, False, False, False, True, False],
 [False, False, False, True, False, True, False, False, True, True],
 [False, False, True, True, True, False, False, False, False, False],
 [False, True, False, False, True, False, True, True, False, False],
 [False, True, False, True, True, True, True, True, True, False],
 [False, False, False, False, True, True, False, True, True, False],
 [True, True, True, True, True, True, False, True, True, True],
 [False, False, True, True, False, False, False, False, True, True],
 [True, True, False, True, True, True, False, True, False, True],
 [True, True, True, True, True, False, False, True, False, False],
 [True, False, False, False, False, True, True, False, True, False],
 [False, True, False, True, True, True, False, False, True, True],
 [True, False, True, False, True, True, True, False, False, False],
 [False, True, False, True, True, True, True, False, False, True],
 [True, False, False, False, True, False, False, Fal

### Histograms of simulated Bernoulli Experiments

Plot the results of our sample distribution simulation.

We count the number of successes in each trial and sort the trials from lowest number of successes to highest.

In [121]:
x = sorted([sum(sim) for sim in simulation1])
x

[3, 3, 4, 4, 4, 4, 4, 5, 5, 6, 6, 6, 7, 7, 9]

(The histnorm parameter normalises the counts to a  between 0 and 1 so that we can easily compare the results from a different number of trials)

In [122]:
import plotly.graph_objects as go
fig = go.Figure()
fig.add_histogram(x=x, name=f"Trials={trials}", histnorm="probability density")

Let's increase the number of trials for a second experiment.

In [0]:
trials2 = 200
simulation2 = bernoulli_sample_distribution(trials=trials2, n=n, p=p)

In [125]:
x2 = sorted([sum(sim) for sim in simulation2])
fig.add_histogram(x=x2, name=f"Trials={trials2}", histnorm="probability density")

And increase a lot for a third.

In [0]:
trials3 = 1000
simulation3 = bernoulli_sample_distribution(trials=trials3, n=n, p=p)

In [127]:
x3 = sorted([sum(sim) for sim in simulation3])
fig.add_histogram(x=x3, name=f"Trials={trials3}", histnorm="probability density")

As we increase the number of trials our simulated binomial distribution begins so resemble the true binomial sampling distribution.

We create the true numbers (where the number of trials approaches infinity) using the scipy-implementation of the binomial distribution. We are plotting the corresponding *probability mass function*.

In [0]:
from scipy.stats import binom

domain = list(range(n))
y_true = binom.pmf(domain, n, p)

In [129]:
fig.add_scatter(x=domain, y=x_true, name="True Binomial")

### Formal description

If a random variable $X$ is follows a binomial disrtibution we write $X \sim B(n,p)$.

That is short for: The random variable $X$ behaves similarly to a Binomial distribution with a $n$ experiments per trial and a sucess probability of $p$.

### Binomial Probability Mass Function

The *probability mass function* returns the probability of getting $k$ successes in $n$ independent bernoulli experiments.

The values above returned by the scipy implementation of the binomial.pmf are:

In [130]:
y_true

array([0.00097656, 0.00976563, 0.04394531, 0.1171875 , 0.20507813,
       0.24609375, 0.20507813, 0.1171875 , 0.04394531, 0.00976563])

In [132]:
list(zip(domain, y_true))

[(0, 0.0009765625),
 (1, 0.00976562500000001),
 (2, 0.04394531249999999),
 (3, 0.11718750000000014),
 (4, 0.20507812500000022),
 (5, 0.24609375000000025),
 (6, 0.20507812500000022),
 (7, 0.11718750000000014),
 (8, 0.04394531249999999),
 (9, 0.00976562500000001)]

So the probability to obtain exactly 3 successes in 10 trials is approx. 0.117.

The pmf can be derived as follows:

- The probability of $k$ successes is $p^k$.
- The probability of $n-k$ failures is $(1-p)^{n-k}$.
- There are ${n\choose k}$ possibilities to distribute k sucesses in a series of n experiments.

Putting all together yields the *binomial probability mass function*:

$f(k, n, p) = Pr(X = k) = {n\choose k} p^k(1-p)^{n-k}$

#### Implementing the probability mass function

We can implement the probability mass function ourselves.

In [0]:
from math import factorial

def binomial_coefficient(n, k):
  return factorial(n) / (factorial(k) * factorial(n-k))

def binomial_pmf(k, n, p):
  return binomial_coefficient(n=n, k=k) * p**k * (1-p)**(n-k)


Veryfing the scipy result for 3 successes (k=3).

In [156]:
binomial_pmf(n=n, k=3, p=p)

0.1171875

### The cumulative distribution function for the binomial distribution

The cumulative distribution function returns the probability of *up to k* successes. Hence it is simply summing up the results of the *probability mass function* up to a value of $k$.

$Pr(X \leq k) = \sum\limits_{i=0}^k {n\choose i} p^i(1-p)^{n-i}$



In [0]:
def binomial_cdf(k, n, p):
  return sum([binomial_pmf(k=i, n=n, p=p) for i in range(k+1)]) #+1 because in python range does not include the endpoint

### 